File: client/web/src/features/vault/api/icon-library-api.ts
import { apiClient } from "@/lib/api-client";

export interface IconLibraryResponse {
  encryptedPayload: string; // base64; empty string when no library exists yet
  version: number;          // server's monotonic version; 0 means no library
  updatedAt: string | null;
}

export interface IconLibraryUpdateRequest {
  encryptedPayload: string;
  expectedVersion: number;
}

export interface IconLibraryUpdateResponse {
  version: number;
  updatedAt: string;
}

export function fetchIconLibrary(): Promise<IconLibraryResponse> {
  return apiClient.get<IconLibraryResponse>("/vault/icon-library");
}

export function putIconLibrary(
  encryptedPayload: string,
  expectedVersion: number,
): Promise<IconLibraryUpdateResponse> {
  return apiClient.put<IconLibraryUpdateResponse>(
    "/vault/icon-library",
    { encryptedPayload, expectedVersion },
  );
}

File: client/android/core/src/main/java/io/github/pstanar/pstotp/core/api/IconLibraryApi.kt
package io.github.pstanar.pstotp.core.api

import org.json.JSONObject

/**
 * Snapshot of the server's icon-library row. `encryptedPayload` is
 * empty and `version` is 0 when the user hasn't written a library yet.
 */
data class IconLibraryResponse(
    val encryptedPayload: String,
    val version: Int,
    val updatedAt: String?,
)

data class IconLibraryUpdateResponse(
    val version: Int,
    val updatedAt: String,
)

class IconLibraryApi(private val client: ApiClient) {

    suspend fun fetch(): IconLibraryResponse =
        client.get("/vault/icon-library") { json ->
            IconLibraryResponse(
                encryptedPayload = json.optString("encryptedPayload", ""),
                version = json.optInt("version", 0),
                updatedAt = json.optString("updatedAt", "").ifEmpty { null },
            )
        }

    suspend fun update(
        encryptedPayload: String,
        expectedVersion: Int,
    ): IconLibraryUpdateResponse {
        val body = JSONObject().apply {
            put("encryptedPayload", encryptedPayload)
            put("expectedVersion", expectedVersion)
        }
        return client.put("/vault/icon-library", body) { json ->
            IconLibraryUpdateResponse(
                version = json.getInt("version"),
                updatedAt = json.getString("updatedAt"),
            )
        }
    }
}
